tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -3844,6 +3844,263 @@ LogicalResult lowerWithScheduleSparseReshapeOpCPU(
+        loc,
+        /* resultTypes */ ArrayRef<Type>({b.getIndexType()}),
+        /* cond */ if_empty_row, /*withElseRegion*/ true);
+
+    if_op.getThenRegion().front().clear();
+    if_op.getElseRegion().front().clear();
+    b.setInsertionPointToStart(&if_op.getThenRegion().front());
+
+    // output_indices
+    SmallVector<Value, 2> store_index{row_start_idx, zero};
+    b.create<memref::StoreOp>(
+        loc, b.create<arith::IndexCastOp>(loc, b.getIntegerType(64), i),
+        output_indices, store_index);
+    store_index[1] = one;
+    b.create<memref::StoreOp>(loc, zero_int, output_indices, store_index);
+    // output_values
+    auto default_value = b.create<memref::LoadOp>(
+        loc, sparse_fill_empty_rows_op.default_value());
+    b.create<memref::StoreOp>(loc, default_value, output_values, row_start_idx);
+    b.create<scf::YieldOp>(loc, one);
+
+    b.setInsertionPointToStart(&if_op.getElseRegion().front());
+    Value row_count = b.create<arith::IndexCastOp>(
+        loc, b.getIndexType(),
+        b.create<memref::LoadOp>(loc, row_count_memref, i));
+    b.create<scf::YieldOp>(loc, row_count);
+
+    b.setInsertionPointToEnd(for_op.getBody());
+    Value yield_value =
+        b.create<arith::AddIOp>(loc, row_start_idx, if_op.getResult(0));
+    b.create<scf::YieldOp>(loc, yield_value);
+    b.setInsertionPointAfter(for_op);
+  }
+
+  // for (i = 0; i < indices.shape(0); ++i) {
+  //   output_idx = reverse_index_map[i]
+  //   output_indices[output_idx, 0] = indices[i, 0]
+  //   output_indices[output_idx, 1] = indices[i, 1]
+  //   output_values[output_idx] = values[i]
+  // }
+  // fill non-empty rows
+  {
+    Value num_indices = b.create<memref::DimOp>(loc, indices, 0);
+    auto for_op = b.create<scf::ForOp>(loc, /* lowerBound */ zero,
+                                       /* upperBound */ num_indices,
+                                       /* step */ one);
+    for_op.getBody()->clear();
+    b.setInsertionPointToStart(for_op.getBody());
+    Value i = for_op.getInductionVar();
+    Value output_idx = b.create<arith::IndexCastOp>(
+        loc, b.getIndexType(),
+        b.create<memref::LoadOp>(loc, reverse_index_map, i));
+    SmallVector<Value, 2> store_index{output_idx, zero};
+    SmallVector<Value, 2> load_index{i, zero};
+    b.create<memref::StoreOp>(
+        loc, b.create<memref::LoadOp>(loc, indices, load_index), output_indices,
+        store_index);
+    store_index[1] = one;
+    load_index[1] = one;
+    b.create<memref::StoreOp>(
+        loc, b.create<memref::LoadOp>(loc, indices, load_index), output_indices,
+        store_index);
+    b.create<memref::StoreOp>(loc, b.create<memref::LoadOp>(loc, values, i),
+                              output_values, output_idx);
+    b.create<scf::YieldOp>(loc, ValueRange{});
+    b.setInsertionPoint(sparse_fill_empty_rows_op.getOperation());
+  }
+
+  // TODO: Support fusion
+  for (Operation* root_op : root_ops) root_op->erase();
+  return success();
```

Comment:
Will there be any performance gain if `output_indeces` of both empty and non-empty are filled in a single for loop pass instead of in to loops ?

Reply:
It is possible to merge 2 loops together, but I am not sure by doing so, how the extra `if-else`s will affect the final performance.